### PR TITLE
Refactor resource selector

### PR DESCRIPTION
**Pull request by kiiadi**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

## Description

Port the `ResourceSelector` to use the new `AwsResourceCache` which allows hiding some of the loading / state concerns - users can just use it like a 'normal' `ComboBox`.

Also show loading and exception details to the customer.

## Related Issue(s)

#726 

## Testing

Updated unit tests for the `ResourceSelector` - still need to figure out which ones can be removed

## Checklist

- [x] I have read the **README** document
- [x] I have read the **CONTRIBUTING** document
- [x] Local run of `gradlew check` succeeds
- [x] My code follows the code style of this project
- [x] I have added tests to cover my changes
- [x] All new and existing tests passed
- [ ] A short description of the change has been added to the **CHANGELOG**

## License
I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/cloudformation/CloudFormationResources.kt [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+import software.amazon.awssdk.services.cloudformation.CloudFormationClient
+import software.amazon.awssdk.services.cloudformation.model.StackStatus
+import software.aws.toolkits.jetbrains.core.ClientBackedCachedResource
+import software.aws.toolkits.jetbrains.core.Resource
+
+object CloudFormationResources {
+    @JvmField
+    val ACTIVE_STACK_SUMMARIES = ClientBackedCachedResource(CloudFormationClient::class, "cfn.active_stacks") {
+        listStackSummariesFilter { it.stackStatus() != StackStatus.DELETE_COMPLETE }
```
Comment (abrooksv) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/cloudformation/CloudFormationResources.kt]: Do we listStackSummariesFilter anywhere else? Not sure if really needs to exist

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/iam/Iam.kt [hunk 2]
```diff
@@ -19,4 +23,19 @@ data class IamRole(val arn: String) {
...
+
+    private val LIST_RAW_ROLES = ClientBackedCachedResource(IamClient::class, "iam.list_roles") {
+        listRolesPaginator().roles().toList()
+    }
+
+    @JvmField
+    val LIST_LAMBDA_ROLES: Resource<List<IamRole>> = Resource.View(LIST_RAW_ROLES) {
+        filter { it.assumeRolePolicyDocument().contains(LAMBDA_PRINCIPAL) }
+            .map { IamRole(it.arn()) }
+            .sortedWith(Comparator.comparing<IamRole, String>(Function { t -> t.toString() }, String.CASE_INSENSITIVE_ORDER))
```
Comment (abrooksv) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/iam/Iam.kt]: The sort feels like an implementation detail that should not live at this layer

File: jetbrains-core/src/software/aws/toolkits/jetbrains/ui/ResourceSelector.kt [hunk 3]
```diff
@@ -3,146 +3,122 @@
...
-            ResourceLoadingStatus.SUCCESSFUL -> ValidationInfo(notSelected, component)
+    }
+
+    private fun processFailure(error: Throwable) {
+        loadingStatus = Status.NOT_LOADED
+        val message = message("loading_resource.failed")
+        LOG.warn(error) { message }
+        runInEdt(ModalityState.any()) {
+            super.setSelectedItem(message)
+            toolTipText = error.message
```
Comment (abrooksv) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/ui/ResourceSelector.kt]: This affects the tooltips we use to indicate what a field does, are there any alternatives?

File: jetbrains-core/tst/software/aws/toolkits/jetbrains/ui/ResourceSelectorTest.kt [hunk 4]
```diff
@@ -4,120 +4,172 @@
...
-
-        comboBox.isEnabled = false
-        comboBox.populateValues(updateStatus = true) { items }
-
-        waitForPopulationComplete(comboBox, items.size)
-        assertThat(comboBox.isEnabled).isEqualTo(true)
-    }
+    fun usePreviouslySelectedItemAfterReloadUnlessSelectItemSet() { }
+
+//    @Test
```
Comment (abrooksv) [inline jetbrains-core/tst/software/aws/toolkits/jetbrains/ui/ResourceSelectorTest.kt]: Fix or delete

File: jetbrains-core/tst/software/aws/toolkits/jetbrains/ui/ResourceSelectorTest.kt [hunk 5]
```diff
@@ -4,120 +4,172 @@
...
-
-        arrayOf("foo", "bar").forEach { comboBox.addItem(it) }
-
-        val exception = Exception("Failed")
-        comboBox.populateValues { throw exception }
-
-        waitForPopulationComplete(comboBox, 0)
-        assertThat(comboBox.loadingException).isEqualTo(exception)
-    }
+//    @Test
```
Comment (abrooksv) [inline jetbrains-core/tst/software/aws/toolkits/jetbrains/ui/ResourceSelectorTest.kt]: Fix or delete

File: jetbrains-core/src/software/aws/toolkits/jetbrains/ui/ResourceSelector.kt [hunk 6]
```diff
@@ -3,146 +3,122 @@
...
-                super.setEnabled(false)
-                false
+    @JvmOverloads
+    @Synchronized
+    fun reload(forceFetch: Boolean = false) {
+        val previouslySelected = actualModel.selectedItem
+        loadingStatus = Status.LOADING
+        runInEdt(ModalityState.any()) {
+            super.setEnabled(false)
+            setEditable(true)
```
Comment (abrooksv) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/ui/ResourceSelector.kt]: Why is it editable when loading?

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/iam/Iam.kt (hunk 2 above)
Comment (kiiadi) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/iam/Iam.kt]: the sorting? yeah we can remove that - the resource selector actually sorts by default

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/cloudformation/CloudFormationResources.kt (hunk 1 above)
Comment (kiiadi) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/cloudformation/CloudFormationResources.kt]: fair

File: jetbrains-core/src/software/aws/toolkits/jetbrains/ui/ResourceSelector.kt (hunk 6 above)
Comment (kiiadi) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/ui/ResourceSelector.kt]: that's the only way we can get it to update `selectedText` you can't set a non-`T` value unless it's editable - it's disabled so the effect isn't actually seen by the customer.

Comment (abrooksv) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/ui/ResourceSelector.kt]: Ah, k

File: jetbrains-core/tst/software/aws/toolkits/jetbrains/ui/ResourceSelectorTest.kt (hunk 4 above)
Comment (kiiadi) [reply jetbrains-core/tst/software/aws/toolkits/jetbrains/ui/ResourceSelectorTest.kt]: done

File: jetbrains-core/tst/software/aws/toolkits/jetbrains/ui/ResourceSelectorTest.kt (hunk 5 above)
Comment (kiiadi) [reply jetbrains-core/tst/software/aws/toolkits/jetbrains/ui/ResourceSelectorTest.kt]: done

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/iam/Iam.kt (hunk 2 above)
Comment (kiiadi) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/iam/Iam.kt]: removed

File: jetbrains-core/src/software/aws/toolkits/jetbrains/ui/ResourceSelector.kt (hunk 3 above)
Comment (kiiadi) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/ui/ResourceSelector.kt]: fixed

File: jetbrains-core/src/software/aws/toolkits/jetbrains/ui/ResourceSelector.kt [hunk 7]
```diff
@@ -3,146 +3,131 @@
...
-class ResourceSelector<T> : ComboBox<T>() {
+typealias Selector<T> = Either<Any?, (T) -> Boolean>
 
-    var loadingStatus: ResourceLoadingStatus = ResourceLoadingStatus.SUCCESSFUL
-        private set
+class ResourceSelector<T>(project: Project, private val resourceType: Resource<out Collection<T>>) : ComboBox<T>() {
 
+    private val resourceCache = AwsResourceCache.getInstance(project)
+    private val actualModel = this.model as DefaultComboBoxModel<T>
+    private var loadingStatus: Status = Status.NOT_LOADED
```
Comment (abrooksv) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/ui/ResourceSelector.kt]: This looks like it is not always accessed on EDT, make volatile?

File: jetbrains-core/src/software/aws/toolkits/jetbrains/ui/ResourceSelector.kt [hunk 8]
```diff
@@ -3,146 +3,131 @@
...
-        val component = if (super.isEnabled()) this else null
-        return when (loadingStatus) {
-            ResourceLoadingStatus.LOADING -> ValidationInfo(loading, component)
-            ResourceLoadingStatus.FAILED -> ValidationInfo(loadingException?.message ?: failed, component)
-            ResourceLoadingStatus.SUCCESSFUL -> ValidationInfo(notSelected, component)
+    }
+
+    private fun processFailure(error: Throwable) {
+        val message = message("loading_resource.failed")
+        LOG.error(error) { message }
```
Comment (abrooksv) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/ui/ResourceSelector.kt]: You want this reported to customer as an IDE error?

Comment (kiiadi) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/ui/ResourceSelector.kt]: yes - I think we should - it represents a state where they can't continue

**Comment by kiiadi** [comment]
Screen Shot 2019-08-06 at 9 31 06 PM

File: jetbrains-core/src/software/aws/toolkits/jetbrains/ui/ResourceSelector.kt [hunk 9]
```diff
@@ -3,146 +3,134 @@
...
-            ResourceLoadingStatus.SUCCESSFUL -> ValidationInfo(notSelected, component)
+    }
+
+    private fun processFailure(error: Throwable) {
+        val message = message("loading_resource.failed")
+        LOG.warn(error) { message }
+        runInEdt(ModalityState.any()) {
+            loadingStatus = Status.NOT_LOADED
+            super.setSelectedItem(message)
+            notifyError(message, error.stackTrace.joinToString("\n"), project)
```
Comment (abrooksv) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/ui/ResourceSelector.kt]: `ExceptionUtil.getThrowableText` makes this cleaner

**Pull request merged**